packages/tools/src/utilities/segmentation/contourAndFindLargestBidirectional.ts

```diff
@@ -0,0 +1,294 @@
+  }
+
+  const segData = vol.imageData.getPointData().getScalars().getData();
+  const width = vol.dimensions[0];
+  const pixelsPerSlice = width * vol.dimensions[1];
+
+  return {
+    /**
+     * Find the center point between point1 and point2, convert it to IJK space
+     * and test if the value at that location is in the segment
+     */
+    testCenter: (point1, point2) => {
+      const point = vec3.add(vec3.create(), point1, point2).map((it) => it / 2);
+      const ijk = vol.imageData.worldToIndex(point as vec3).map(Math.round);
+      const [i, j, k] = ijk;
+      const index = i + j * width + k * pixelsPerSlice;
+      const value = segData[index];
+      return value === segmentIndex || containedSegmentIndices?.has(value);
+    },
+
+    toIJK: (point) => vol.imageData.worldToIndex(point as vec3),
+
+    testIJK: (ijk) => {
+      const [i, j, k] = ijk;
+      const index =
+        Math.round(i) + Math.round(j) * width + Math.round(k) * pixelsPerSlice;
+      const value = segData[index];
+      return value === segmentIndex || containedSegmentIndices?.has(value);
+    },
+  };
+}
+
+/**
+ * Search in the contours for the given segment to find the largest bidirectional
+ * that will fit entirely within the slice contours inside the contours object.
+ * Assumptions/implementation details:
+ *
+ *   1. The major and minor bidirectional lines must not cross the contour
+ *   2. The center point for both major and minor bidirectional lines must be
+ *      within the segment, or the contained segment index.
+ *   3. The major/minor axis must be orthogonal
+ *
+ * Note this does NOT test that the major/minor axis intersect.  Normally they will, but
+ * it isn't a hard requirement.
+ *
+ * The way that islands within the contours are handled is to allow the island to be
+ * coloured with something that is contained - that way both open and closed islands
+ * can be handled correctly for finding the bidirectional  (an open island is a section
+ * inside the segment that is open to the outside - this can happen at bone endpoints or when
+ * one region flows into another)
+ */
+function findLargestBidirectional(contours, segVolumeId: string, segment) {
+  const { sliceContours } = contours;
+  const { segmentIndex, containedSegmentIndices } = segment;
+  let maxBidirectional;
+  const isInSegment = createIsInSegment(
+    segVolumeId,
+    segmentIndex,
+    containedSegmentIndices
+  );
+  for (const sliceContour of sliceContours) {
+    const bidirectional = createBidirectionalForSlice(
+      sliceContour,
+      isInSegment,
+      maxBidirectional
+    );
+    if (!bidirectional) {
+      continue;
+    }
+    maxBidirectional = bidirectional;
+  }
+  if (maxBidirectional) {
+    Object.assign(maxBidirectional, segment);
+  }
+  return maxBidirectional;
+}
+
+const lengthSq = (point) =>
+  point[0] * point[0] + point[1] * point[1] + point[2] * point[2];
+
+const distanceSq = (p1, p2) => lengthSq(vec3.sub(vec3.create(), p1, p2));
```

Comment:
we already have these utilities, please use them

Reply:
What are they called?  I couldn't find them.

Reply:
Found them.